map_find.h

```diff
@@ -0,0 +1,150 @@
+class MapFind
+{
+public:
+    typedef typename Map::key_type KeyType;
+    typedef typename Map::mapped_type ValueType;
+    typedef std::reference_wrapper<const ValueType> ConstRefValueType;
+
+    /**
+     * @brief Create a MapFind object on a map, no find is performed until find() is called.
+     *
+     * @param map : the map to perform the search on
+     */
+    MapFind(Map & map): map(map) {}
+
+
+    /**
+     * @brief Create a MapFind object and find the key in the associated map
+     *
+     * @param map : the map to perform the search on
+     * @param key : the key to be searched in the map
+     */
+    MapFind(Map & map, const KeyType & key): map(map) {
+        find(key);
+    }
+
+    /**
+     * @brief Callback fn with the value associated to the searched key
+     * if the find was successful
+     *
+     * @param fn : a callback function that takes a constant mapped type reference as parameter
+     * @returns A MapFind object to query the restult or perfom another find
+     */
+    MapFind& if_found(std::function<void(const ValueType &)> fn) {
+        if( res ) {
+            fn( res->get() );
+        }
+        return *this;
+    }
+
+    /**
+     * @brief Callback fn if nothing was found in the map
+     *
+     * @param fn : a callback function with no parameter
+     * @returns A MapFind object to query the restult or perfom another find
+     */
+    template<class Func>
+    MapFind& if_not_found(Func fn) {
+        if( !res ) {
+            fn();
+        }
+        return *this;
+    }
+
+    /**
+     * @brief Search for key within the map used in the constructor
+     *
+     * @param key : the key to be searched in the map
+     * @returns A MapFind object to query the restult or perfom another find
+     */
+    template<class Key>
+    MapFind& find(const Key & key) {
+        auto it = map.find(key);
+        if(it != map.end()) {
+            auto cref_val = std::cref(it->second);
+            res = boost::make_optional(cref_val);
+        } else {
+            res = boost::none;
+        }
+        return *this;
+    }
+
+private:
+    Map & map;
+    boost::optional<ConstRefValueType> res;
+};
+
+/**
+ * @brief Create a MapFind object with the associated map
+ *
+ * This function facilitates the contruction of a MapFind
+ * as it allowes type deduction.
+ * @param map : the map to perform the search on
+ * @returns A MapFind object to query the restult or perfom another find
+ */
+template<class Map>
+MapFind<Map> make_map_find(Map& map) {
+    return MapFind<Map>(map);
+}
+
+/**
+ * @brief Create a MapFind object with the associated map
+ *
+ * This function facilitates the contruction of a MapFind
+ * as it allowes type deduction.
+ * @param map : the map to perform the search on
+ * @param key : the key to search in the map
+ * @returns A MapFind object to query the restult or perfom another find
+ */
+template<class Map, class Key>
+MapFind<Map> make_map_find(Map& c, const Key& key) {
```

Comment:
how about add a default value if not found?

Reply:
you already have something for this in `functions.h`
```cpp
template<typename Map>
const typename Map::mapped_type&
find_or_default(const typename Map::key_type& k, const Map& m);
```